src/protocol/sort/multi_bit_permutation.rs

```diff
@@ -0,0 +1,256 @@
+/// At a high level, the protocol works as follows:
+/// 1. Start with a vector of list of `L*n` secret shares `[[x1_1 ... x1_n], .. , [xL_1 ... xL_n]]` where each is a secret sharing of either zero or one.
+///    Here, L is the number of multi bits which are procrssed together (`num_multi_bits`) and n is the number of records
+/// 2. Equality Bit Checker : For j in 0 to 2 pow `num_multi_bits`
+///    i. Get binary representation of j (B1 .. BL)
+///    ii. For i in `num_multi_bits`
+///      a. Locally compute `mult_inputs` as (Bi * `xi_j` + (1-Bi)(1- `xi_j`))
+///   iii. Multiply all `mult_inputs` for this j
+/// 3. Compute accumulated sum: For j in 0 to 2 pow `num_multi_bits`
+///    i. For each record
+///       a. Calculate accumulated `total_sum` = s + `mult_output`
+/// 4. Compute the final output using sum of products executed in parallel for each record.
+#[allow(dead_code)]
+pub async fn multi_bit_permutation<'a, F: Field, S: SecretSharing<F>, C: Context<F, Share = S>>(
+    ctx: C,
+    input: &[Vec<S>],
+) -> Result<Vec<S>, Error> {
+    let num_multi_bits = input.len();
+    let num_records = input[0].len();
+    let num_possible_bit_values = 2 << (num_multi_bits - 1);
+
+    // Equality bit checker: this checks if each secret shared record is equal to any of numbers between 0 and num_possible_bit_values
+    let mut equality_check_futures = Vec::with_capacity(num_possible_bit_values * num_records);
+    for j in 0..num_possible_bit_values {
+        let ctx_equality_checker = ctx.narrow(&EqualityBitChecker);
+        equality_check_futures
+            .push(async move { get_bit_equality_checkers(j, input, ctx_equality_checker).await });
+    }
+
+    let equality_checks = try_join_all(equality_check_futures).await?;
+
+    // Compute accumulated sum
+    let mut sop_inputs_transposed = Vec::with_capacity(num_possible_bit_values * num_records);
+    let mut initial_state = S::ZERO;
+    for equality_check in equality_checks {
+        let out = equality_check
+            .into_iter()
+            .scan(initial_state, |sum, x| {
+                *sum += &x;
+                Some((x, sum.clone()))
+            })
+            .collect::<Vec<_>>();
+        initial_state = out
+            .last()
+            .unwrap_or_else(|| panic!("We should never reach here"))
+            .1
+            .clone();
+        sop_inputs_transposed.push(out);
+    }
+
+    // Take sum of products of output of equality check and accumulated sum
+    let mut permutation_futures = Vec::with_capacity(num_records);
+    for rec in 0..num_records {
+        let ctx_sop = ctx.narrow(&Sop);
+        let mut sop_inputs = Vec::with_capacity(num_possible_bit_values);
+        for sop_input_transposed in sop_inputs_transposed.iter().take(num_possible_bit_values) {
+            sop_inputs.push((&sop_input_transposed[rec].0, &sop_input_transposed[rec].1));
+        }
```

Comment:
It seems that in the current implementation `sop_inputs_transposed` contains a copy of the original "equality_checks", which are subsequently copied again here. That seems wasteful. I'd recommend not generating that intermediate copy. You can still construct this vector of pairs here from the original `equality_checks` and the cumulative sums.